### PR TITLE
to_chars for long double and binary128

### PR DESCRIPTION
**Pull request by mborland**
Derivative of ryu_generic_128 methods.

### CODE REVIEW

**Comment by mborland** [comment]
@pdimov Do you have any idea why GCC 4.8 would have an issue with signaling NaN? The same issue occurred when I used the glibc issignaling macro. I didn't see anything in the 4.9 release notes that would explain the error.

```
====== BEGIN OUTPUT ======
libs/charconv/test/to_chars_float.cpp(86): test 'buffer5 == "nan(snan)"' ('nan' == 'nan(snan)') failed in function 'void non_finite_values(boost::charconv::chars_format, int) [with T = long double]'
libs/charconv/test/to_chars_float.cpp(92): test 'buffer6 == "-nan(snan)"' ('-nan(ind)' == '-nan(snan)') failed in function 'void non_finite_values(boost::charconv::chars_format, int) [with T = long double]'
2 errors detected.

EXIT STATUS: 2
====== END OUTPUT ======
```

**Comment by mborland** [comment]
```
====== BEGIN OUTPUT ======
GNU C++ version 14.0.0 20230612 (experimental)
GNU libstdc++ version 20230612

            std::snprintf<float>, scientific, 0:  4071 ms (s=1249494190)
            std::to_chars<float>, scientific, 0:   547 ms (s=1222675070)
boost::charconv::to_chars<float>, scientific, 0:   435 ms (s=1222675070)

            std::snprintf<float>, scientific, 6:  3902 ms (s=1209648310)
            std::to_chars<float>, scientific, 6:  1047 ms (s=1209648310)
boost::charconv::to_chars<float>, scientific, 6:   458 ms (s=1209648310)

            std::snprintf<float>, general, 0:  4065 ms (s=1215305890)
            std::to_chars<float>, general, 0:   632 ms (s=1213059940)
boost::charconv::to_chars<float>, general, 0:   480 ms (s=1212721020)

            std::snprintf<float>, general, 6:  4051 ms (s=1177644030)
            std::to_chars<float>, general, 6:  1187 ms (s=1177644030)
boost::charconv::to_chars<float>, general, 6:   483 ms (s=1207425580)

            std::snprintf<double>, scientific, 0:  7615 ms (s=1427249960)
            std::to_chars<double>, scientific, 0:   499 ms (s=1458911385)
boost::charconv::to_chars<double>, scientific, 0:   507 ms (s=1415002950)

            std::snprintf<double>, scientific, 6:  5599 ms (s=1227341370)
            std::to_chars<double>, scientific, 6:  1101 ms (s=1227341370)
boost::charconv::to_chars<double>, scientific, 6:   488 ms (s=1227341370)

            std::snprintf<double>, general, 0:  7400 ms (s=1402505130)
            std::to_chars<double>, general, 0:   506 ms (s=1448519075)
boost::charconv::to_chars<double>, general, 0:   531 ms (s=1412564790)

            std::snprintf<double>, general, 6:  5379 ms (s=1203912950)
            std::to_chars<double>, general, 6:  1156 ms (s=1203912950)
boost::charconv::to_chars<double>, general, 6:   514 ms (s=1225124730)

            std::to_chars<_Float128>, scientific, 0:  2062 ms (s=5039881560)
boost::charconv::to_chars<_Float128>, scientific, 0:  1810 ms (s=1319968980)

            std::to_chars<_Float128>, scientific, 6: 53017 ms (s=1249958800)
boost::charconv::to_chars<_Float128>, scientific, 6:   319 ms (s=4779887670)

            std::to_chars<_Float128>, general, 0:  2075 ms (s=419990130)
boost::charconv::to_chars<_Float128>, general, 0:  1803 ms (s=1319968980)

            std::to_chars<_Float128>, general, 6: 51577 ms (s=1227660290)
boost::charconv::to_chars<_Float128>, general, 6:   328 ms (s=4779887670)
```

Updated benchmarks with this PR.

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,6 @@ if(HAVE_BOOST_TEST)
 # https://crascit.com/2015/03/28/enabling-cxx11-in-cmake/
 set(CMAKE_CXX_EXTENSIONS OFF)
 add_compile_definitions(BOOST_CHARCONV_CMAKE_TESTING)
-boost_test_jamfile(FILE Jamfile LINK_LIBRARIES Boost::charconv Boost::core Boost::assert)
+boost_test_jamfile(FILE Jamfile LINK_LIBRARIES Boost::charconv Boost::core Boost::assert COMPILE_DEFINITIONS BOOST_CHARCONV_CMAKE_TESTING=1)
```
Comment (mborland) [inline test/CMakeLists.txt]: @pdimov What is the correct way to pass a define to `boost_test_jamfile`? In config I added that if `BOOST_CHARCONV_CMAKE_TESTING` is defined to not use `<quadmath.h>`. The issue with `set(CMAKE_CXX_EXTENSIONS OFF)` not doing what it is supposed to was fixed in CMake 3.23 which I think is too new to require.

Comment (pdimov) [reply test/CMakeLists.txt]: COMPILE_DEFINITIONS is the right way.

**Comment by mborland** [comment]
@pdimov I believe this is good for review when you have time.

File: include/boost/charconv/detail/config.hpp [hunk 2]
```diff
@@ -157,6 +162,20 @@ static_assert((BOOST_CHARCONV_ENDIAN_BIG_BYTE || BOOST_CHARCONV_ENDIAN_LITTLE_BY
 #  define BOOST_CHARCONV_NO_CONSTEXPR_DETECTION
 #endif
 
+#if BOOST_MSVC
```
Comment (pdimov) [inline include/boost/charconv/detail/config.hpp]: This should be `defined(BOOST_MSVC)`.

File: include/boost/charconv/detail/config.hpp [hunk 3]
```diff
@@ -157,6 +162,20 @@ static_assert((BOOST_CHARCONV_ENDIAN_BIG_BYTE || BOOST_CHARCONV_ENDIAN_LITTLE_BY
...
+#  define BOOST_CHARCONV_ASSUME(expr) __assume(expr)
+#elif defined(__clang__)
+#  define BOOST_CHARCONV_ASSUME(expr) __builtin_assume(expr)
+#elif defined(__GNUC__)
+#  define BOOST_CHARCONV_ASSUME(expr) if (expr) {} else { __builtin_unreachable(); }
+#elif defined(__has_cpp_attribute)
+#  if __has_cpp_attribute(assume)
+#    define BOOST_CHARCONV_ASSUME(expr) [[assume(expr)]]
+#  endif
```
Comment (pdimov) [inline include/boost/charconv/detail/config.hpp:176]: If `__has_cpp_attribute(assume)` is 0, `BOOST_CHARCONV_ASSUME` remains undefined.

File: src/to_chars.cpp [hunk 4]
```diff
@@ -66,14 +68,7 @@ namespace boost { namespace charconv { namespace detail { namespace to_chars_det
...
     {
-        BOOST_IF_CONSTEXPR (('0' & 0xf) == 0)
-        {
-            *buffer = char('0' | n);
-        }
-        else
-        {
-            *buffer = char('0' + n);
-        }
+        *buffer = char('0' | n);
```
Comment (pdimov) [inline src/to_chars.cpp]: `'0' | n` is not guaranteed to be valid under all encodings, but `'0' + n` is.

'0'..'9' are guaranteed to be consecutive, but the low hex digit is not guaranteed to be 0.

**Pull request merged**